python/tank/descriptor/io_descriptor/base.py

```diff
@@ -211,93 +212,85 @@ def _find_latest_tag_by_pattern(self, version_numbers, pattern):
-
-        # split our pattern, beware each part is a string (even integers)
-        version_split = re.findall("([0-9]+|x)", pattern)
-        if 'x' in version_split:
-            # check that we don't have an incorrect pattern using x
-            # then a digit, eg. v4.x.2
-            if re.match("^v[0-9\.]+[x\.]+[0-9\.]+$", pattern):
-                raise TankDescriptorError(
-                    "Incorrect version pattern '%s'. "
-                    "There should be no digit after a 'x'." % pattern
-                )
-
-        current = versions
-        version_to_use = None
-        # process each digit in the pattern
-        for version_digit in version_split:
-            if version_digit == 'x':
-                # replace the 'x' by the latest at this level
-                version_digit = max(current.keys(), key=int)
-            version_digit = int(version_digit)
-            if version_digit not in current:
-                # no matches
+        # Split the pattern either with "." or "-", keep the separator in the result
+        # with the grouping in the regexp
+        version_split = re.split("(\.|-)", pattern)
+        num_tokens = len(version_split)
+        # Special case for the first token for which the wildcard is "vx" and not
+        # "x"
+        has_v_prefix = False
+        version_regexp = version_split[0]
+        if version_split[0] == "vx":
+            has_v_prefix = True
+            version_split[0] = "x"
+        # Build a base version string that versions will have to match. The base
+        # version contains all tokens until we find a "x".
+        for i in range(0, num_tokens, 2):
+            token = version_split[i]
+            if token == "x":
+                # if we found a "x", check that all following tokens are "x" as
+                # well, having something like v4.x.2 is illegal
+                # we can't use 'any' here until supporting Python 2.4 is deprecated
+                for token2 in version_split[i+2:num_tokens:2]:
+                    if token2 != "x":
+                        raise TankDescriptorError(
+                            "Incorrect version pattern '%s'. "
+                            "There should be no digit after a 'x', found %s." % (
+                                pattern, token2,
+                            )
+                        )
+                # Build our base version string. Please note we will have a "." or a "-"
+                # at the end.
+                # Please note that we don't have to worry about leading "v" here: if
+                # the pattern starts with "vx" the base version will be an empty
+                # string
+                base_version = "".join(version_split[:i])
+                break
+        else:
+            # No "x" in our pattern
+            base_version = "".join(version_split)
+
+        if not base_version: # Match anything
```

Comment:
This is not the same case. This happens if `vx.x.x` was used as pattern. We need to go through the `x` validation code, to check that we don't have things like `vx.x.3`, so I'm not sure it is possible to combine it with the `no pattern` case?